src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java

```diff
@@ -179,13 +258,171 @@ public void testCreatePileupForReadAndOffset1(){
+        final PileupElement pe = PileupElement.createPileupForReadAndOffset(read, 5);
+        Assert.assertFalse(pe.atStartOfCurrentCigar(), "atStartOfCurrentCigar");
+        Assert.assertFalse(pe.atEndOfCurrentCigar(), "atEndOfCurrentCigar");
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyAfter(op), "isImmediatelyAfter " + op);
+        }
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyBefore(op), "isImmediatelyBefore " + op);
+        }
+    }
+
+    @Test
+    public void testIsImmediatelyAfter_afterLastM() {
+        final GATKRead read = ArtificialReadUtils.createArtificialRead("10M10D10M");
+        final PileupElement pe = PileupElement.createPileupForReadAndOffset(read, 9);
+        Assert.assertFalse(pe.atStartOfCurrentCigar(), "atStartOfCurrentCigar");
+        Assert.assertTrue(pe.atEndOfCurrentCigar(), "atEndOfCurrentCigar");
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyAfter(op), "isImmediatelyAfter " + op);
+        }
+
+        Assert.assertTrue(pe.isImmediatelyBefore(D));
+        for (final CigarOperator op : CigarOperator.values()) {
+            if (op != D) {
+                Assert.assertFalse(pe.isImmediatelyBefore(op), "isImmediatelyBefore " + op);
+            }
+        }
+    }
+
+    @Test
+    public void testIsImmediatelyAfter_afterFirstD() {
+        final GATKRead read = ArtificialReadUtils.createArtificialRead("10M10D10M");
+        final PileupElement pe = PileupElement.createPileupForReadAndOffset(read, 10);
+        Assert.assertFalse(pe.atEndOfCurrentCigar(), "atEndOfCurrentCigar");
+        Assert.assertTrue(pe.isImmediatelyAfter(D));
+        for (final CigarOperator op : CigarOperator.values()) {
+            if (op != D) {
+                Assert.assertFalse(pe.isImmediatelyAfter(op), "isImmediatelyAfter " + op);
+            }
+        }
+
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyBefore(op));
+        }
+    }
+
+    @Test
+    public void testIsImmediatelyAfter_insideDs() {
+        final GATKRead read = ArtificialReadUtils.createArtificialRead("10M10D10M");
+        final PileupElement pe = PileupElement.createPileupForReadAndOffset(read, 15);
+        Assert.assertFalse(pe.atStartOfCurrentCigar(), "atStartOfCurrentCigar");
+        Assert.assertFalse(pe.atEndOfCurrentCigar(), "atEndOfCurrentCigar");
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyAfter(op), "isImmediatelyAfter " + op);
+        }
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyBefore(op), "isImmediatelyBefore " + op);
+        }
+    }
+
+    @DataProvider(name = "elementData")
+    private Object[][] elementData() {
```

Comment:
Use a more descriptive name for your `DataProvider` -- eg., `adjacentElementTestData` or something similar